core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/columns/DataColumnImpl.kt

```diff
@@ -12,6 +17,32 @@ internal abstract class DataColumnImpl<T>(
 ) : DataColumn<T>,
     DataColumnInternal<T> {
 
+    private infix fun <T> T?.matches(type: KType) =
+        when {
+            this == null -> type.isMarkedNullable
+
+            this.isPrimitiveArray ->
+                type.isPrimitiveArray &&
+                    this!!::class.qualifiedName == type.classifier?.let { (it as KClass<*>).qualifiedName }
+
+            this.isArray -> type.isArray
+
+            // cannot check the precise type of array
+            else -> this!!::class.isSubclassOf(type.classifier as KClass<*>)
+        }
+
+    init {
+        /* Check for [Issue #713](https://github.com/Kotlin/dataframe/issues/713).
+         * This only runs with `kotlin.dataframe.debug=true` in gradle.properties
+         */
+        if (BuildConfig.DEBUG) {
+            require(values.all { it matches type }) {
```

Comment:
It checks all values, right? Is it worth to check random 10-20 numbers from the column?

Reply:
If we only check a handful of values there's a chance we won't catch the outlier and get flaky tests.